server/backend/api/app/crud/crud_user.py

```diff
@@ -76,40 +89,59 @@ def create_user(fam_user: schemas.FamUser, db: Session):
 
 
 def find_or_create(
-        db: Session,
-        user_type_code: str,
-        user_name: str,
-        requester: str):
+    db: Session, user_type_code: str, user_name: str, user_guid: str, requester: str
+):
     LOGGER.debug(
-        f"User - 'find_or_create' with user_type: {user_type_code}, " +
-        f"user_name: {user_name}."
+        f"User - 'find_or_create' with user_type: {user_type_code}, "
+        + f"user_name: {user_name}."
     )
 
-    fam_user = get_user_by_domain_and_name(db, user_type_code, user_name)
+    fam_user = get_user_by_domain_and_guid(db, user_type_code, user_guid)
     if not fam_user:
-        request_user = schemas.FamUser(
-            **{
-                "user_type_code": user_type_code,
-                "user_name": user_name,
-                "create_user": requester,
-            }
+        fam_user_by_domain_and_name = get_user_by_domain_and_name(
+            db, user_type_code, user_name
         )
-        fam_user = create_user(request_user, db)
-        LOGGER.debug(f"User created: {fam_user.user_id}.")
-        return fam_user
+        # if not found user by user_guid, but found user by domain and name
+        # in the case of historical FAM user that has no user_guid stored
+        # add their user_guid
+        if fam_user_by_domain_and_name and not fam_user_by_domain_and_name.user_guid:
+            update(
+                db,
+                fam_user_by_domain_and_name.user_id,
+                {models.FamUser.user_guid: user_guid},
+                requester,
+            )
+            LOGGER.debug(
+                f"User {fam_user_by_domain_and_name.user_id} found, updated to store their user_guid."
+            )
+            return fam_user_by_domain_and_name
+        # if not found user by user_guid, and not found user by domain and username,
+        # or found user by domain and username, but user guid does not match (this is the edge case that could happen when username changed from IDP provider)
```

Comment:
I wonder probably no need to add line 119 comment. It will be a new record in database to be inserted anyway (not found by user_guid and not found by user_name).

When IDP change "the same user's" username, is the user_guid also get changed? My assumption is no, but that might be a question to IDP...

Reply:
user_guid will never change I believe

Reply:
so the comment on 119, I mean we found the user by domain and username (USER A) in our database, however this username is now used by another person (USER B), not the original person anymore, that's why the user_guid is not the same. This could happen that we initially grant USER A some access, so we have record for A. Later on, A changed his username, but never login to FAM, so our record is old. And then USER B changed his name to A's old name. When we grant B access, we got A's old name with B's user_guid, in this case, we should add a new user record for USER B. And once USER A login to FAM or any apps integrated with FAM, we'll update A's username.